File: packages/flutterfire_gen/test/helper/app_user_post.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'app_user_post.dart';

class ReadAppUserPost {
  const ReadAppUserPost({
    required this.content,
    required this.numbers,
    required this.updatedAt,
    required this.appUserPostId,
    required this.path,
    required this.appUserPostReference,
  });

  final String content;

  final List<int> numbers;

  final DateTime? updatedAt;

  final String appUserPostId;

  final String path;

  final DocumentReference<ReadAppUserPost> appUserPostReference;

  factory ReadAppUserPost.fromJson(Map<String, dynamic> json) {
    final extendedJson = <String, dynamic>{
      ...json,
    };
    return ReadAppUserPost(
      content: extendedJson['content'] as String? ?? '',
      numbers: (extendedJson['numbers'] as List<dynamic>?)
              ?.map((e) => e as int)
              .toList() ??
          const <int>[],
      updatedAt: (extendedJson['updatedAt'] as Timestamp?)?.toDate(),
      appUserPostId: extendedJson['appUserPostId'] as String,
      path: extendedJson['path'] as String,
      appUserPostReference: extendedJson['appUserPostReference']
          as DocumentReference<ReadAppUserPost>,
    );
  }

  factory ReadAppUserPost.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadAppUserPost.fromJson(<String, dynamic>{
      ...data,
      'appUserPostId': ds.id,
      'path': ds.reference.path,
      'appUserPostReference': ds.reference.parent.doc(ds.id).withConverter(
            fromFirestore: (ds, _) => ReadAppUserPost.fromDocumentSnapshot(ds),
            toFirestore: (obj, _) => throw UnimplementedError(),
          ),
    });
  }

  ReadAppUserPost copyWith({
    String? content,
    List<int>? numbers,
    DateTime? updatedAt,
    String? appUserPostId,
    String? path,
    DocumentReference<ReadAppUserPost>? appUserPostReference,
  }) {
    return ReadAppUserPost(
      content: content ?? this.content,
      numbers: numbers ?? this.numbers,
      updatedAt: updatedAt ?? this.updatedAt,
      appUserPostId: appUserPostId ?? this.appUserPostId,
      path: path ?? this.path,
      appUserPostReference: appUserPostReference ?? this.appUserPostReference,
    );
  }
}

class CreateAppUserPost {
  const CreateAppUserPost({
    required this.content,
    required this.numbers,
  });

  final String content;
  final FirestoreData<List<int>> numbers;

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'content': content,
      'numbers': numbers.value,
      'updatedAt': FieldValue.serverTimestamp(),
    };
    final jsonPostProcessors = <({String key, dynamic value})>[];
    return {
      ...json,
      ...Map.fromEntries(jsonPostProcessors
          .map((record) => MapEntry(record.key, record.value))),
    };
  }
}

class UpdateAppUserPost {
  const UpdateAppUserPost({
    this.content,
    this.numbers,
  });

  final String? content;
  final FirestoreData<List<int>>? numbers;

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      if (content != null) 'content': content,
      if (numbers != null) 'numbers': numbers!.value,
      'updatedAt': FieldValue.serverTimestamp(),
    };
    final jsonPostProcessors = <({String key, dynamic value})>[];
    return {
      ...json,
      ...Map.fromEntries(jsonPostProcessors
          .map((record) => MapEntry(record.key, record.value))),
    };
  }
}

class DeleteAppUserPost {}

/// Provides a reference to the appUserPosts collection for reading.
CollectionReference<ReadAppUserPost> readAppUserPostCollectionReference({
  required String appUserId,
}) =>
    FirebaseFirestore.instance
        .collection('appUsers')
        .doc(appUserId)
        .collection('appUserPosts')
        .withConverter<ReadAppUserPost>(
          fromFirestore: (ds, _) => ReadAppUserPost.fromDocumentSnapshot(ds),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a appUserPost document for reading.
DocumentReference<ReadAppUserPost> readAppUserPostDocumentReference({
  required String appUserId,
  required String appUserPostId,
}) =>
    readAppUserPostCollectionReference(appUserId: appUserId).doc(appUserPostId);

/// Provides a reference to the appUserPosts collection for creating.
CollectionReference<CreateAppUserPost> createAppUserPostCollectionReference({
  required String appUserId,
}) =>
    FirebaseFirestore.instance
        .collection('appUsers')
        .doc(appUserId)
        .collection('appUserPosts')
        .withConverter<CreateAppUserPost>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a appUserPost document for creating.
DocumentReference<CreateAppUserPost> createAppUserPostDocumentReference({
  required String appUserId,
  required String appUserPostId,
}) =>
    createAppUserPostCollectionReference(appUserId: appUserId)
        .doc(appUserPostId);

/// Provides a reference to the appUserPosts collection for updating.
CollectionReference<UpdateAppUserPost> updateAppUserPostCollectionReference({
  required String appUserId,
}) =>
    FirebaseFirestore.instance
        .collection('appUsers')
        .doc(appUserId)
        .collection('appUserPosts')
        .withConverter<UpdateAppUserPost>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a appUserPost document for updating.
DocumentReference<UpdateAppUserPost> updateAppUserPostDocumentReference({
  required String appUserId,
  required String appUserPostId,
}) =>
    updateAppUserPostCollectionReference(appUserId: appUserId)
        .doc(appUserPostId);

/// Provides a reference to the appUserPosts collection for deleting.
CollectionReference<DeleteAppUserPost> deleteAppUserPostCollectionReference({
  required String appUserId,
}) =>
    FirebaseFirestore.instance
        .collection('appUsers')
        .doc(appUserId)
        .collection('appUserPosts')
        .withConverter<DeleteAppUserPost>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a appUserPost document for deleting.
DocumentReference<DeleteAppUserPost> deleteAppUserPostDocumentReference({
  required String appUserId,
  required String appUserPostId,
}) =>
    deleteAppUserPostCollectionReference(appUserId: appUserId)
        .doc(appUserPostId);

/// Manages queries against the appUserPosts collection.
class AppUserPostQuery {
  /// Fetches [ReadAppUserPost] documents.
  Future<List<ReadAppUserPost>> fetchDocuments({
    required String appUserId,
    GetOptions? options,
    Query<ReadAppUserPost>? Function(Query<ReadAppUserPost> query)?
        queryBuilder,
    int Function(ReadAppUserPost lhs, ReadAppUserPost rhs)? compare,
  }) async {
    Query<ReadAppUserPost> query =
        readAppUserPostCollectionReference(appUserId: appUserId);
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [AppUserPost] documents.
  Stream<List<ReadAppUserPost>> subscribeDocuments({
    required String appUserId,
    Query<ReadAppUserPost>? Function(Query<ReadAppUserPost> query)?
        queryBuilder,
    int Function(ReadAppUserPost lhs, ReadAppUserPost rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadAppUserPost> query =
        readAppUserPostCollectionReference(appUserId: appUserId);
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadAppUserPost] document.
  Future<ReadAppUserPost?> fetchDocument({
    required String appUserId,
    required String appUserPostId,
    GetOptions? options,
  }) async {
    final ds = await readAppUserPostDocumentReference(
      appUserId: appUserId,
      appUserPostId: appUserPostId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [AppUserPost] document.
  Stream<ReadAppUserPost?> subscribeDocument({
    required String appUserId,
    required String appUserPostId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readAppUserPostDocumentReference(
      appUserId: appUserId,
      appUserPostId: appUserPostId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [AppUserPost] document.
  Future<DocumentReference<CreateAppUserPost>> add({
    required String appUserId,
    required CreateAppUserPost createAppUserPost,
  }) =>
      createAppUserPostCollectionReference(appUserId: appUserId)
          .add(createAppUserPost);

  /// Sets a [AppUserPost] document.
  Future<void> set({
    required String appUserId,
    required String appUserPostId,
    required CreateAppUserPost createAppUserPost,
    SetOptions? options,
  }) =>
      createAppUserPostDocumentReference(
        appUserId: appUserId,
        appUserPostId: appUserPostId,
      ).set(createAppUserPost, options);

  /// Updates a specific [AppUserPost] document.
  Future<void> update({
    required String appUserId,
    required String appUserPostId,
    required UpdateAppUserPost updateAppUserPost,
  }) =>
      updateAppUserPostDocumentReference(
        appUserId: appUserId,
        appUserPostId: appUserPostId,
      ).update(updateAppUserPost.toJson());

  /// Deletes a specific [AppUserPost] document.
  Future<void> delete({
    required String appUserId,
    required String appUserPostId,
  }) =>
      deleteAppUserPostDocumentReference(
        appUserId: appUserId,
        appUserPostId: appUserPostId,
      ).delete();
}

File: packages/flutterfire_gen_ts/test/helper/app_user_post.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    DocumentReference,
    FieldValue,
    QueryDocumentSnapshot,
    QuerySnapshot,
    WriteResult
} from 'firebase-admin/firestore'

export class ReadAppUserPost {
  constructor ({
    appUserPostId,
    path,
    content,
numbers,
updatedAt,
  }: {
    appUserPostId: string
    path: string
    content: string
numbers: number[]
updatedAt?: Date
  }) {
    this.appUserPostId = appUserPostId
    this.path = path
    this.content = content
this.numbers = numbers
this.updatedAt = updatedAt
  }

  readonly appUserPostId: string

  readonly path: string

  readonly content: string

readonly numbers: number[]

readonly updatedAt?: Date

  private static fromJson(json: Record<string, unknown>): ReadAppUserPost {
    return new ReadAppUserPost({
      appUserPostId: json['appUserPostId'] as string,
      path: json['path'] as string,
      content: (json['content'] as string | undefined) ?? '',
numbers: (json['numbers'] as (unknown[]  | undefined))?.map((e) => e as number).toList() ?? const <int>[],
updatedAt: (json['updatedAt'] as FirebaseFirestore.Timestamp | undefined)?.toDate(),
    })
  }


  static fromDocumentSnapshot(
    ds: FirebaseFirestore.DocumentSnapshot
  ): ReadAppUserPost {
      const data = ds.data()!
      const cleanedData: Record<string, unknown> = {}
      for (const [key, value] of Object.entries(data)) {
          cleanedData[key] = value === null ? undefined : value
      }
      return ReadAppUserPost.fromJson({
          ...cleanedData,
          appUserPostId: ds.id,
          path: ds.ref.path
      })
  }

}

export class CreateAppUserPost {
  constructor({
    content,
numbers,
updatedAt,
  }: {
    content: string
numbers: number[] | FieldValue
updatedAt?: Date
  }) {
    this.content = content
this.numbers = numbers
this.updatedAt = updatedAt
  }


  readonly content: string

readonly numbers: number[] | FieldValue

readonly updatedAt?: Date

  toJson(): Record<string, unknown> {
  return {
    content: this.content,
numbers: this.numbers,
updatedAt: this.updatedAt,
  };
}

}

export class UpdateAppUserPost {
  constructor({
    content,
numbers,
updatedAt,
  }: {
    content?: string
numbers?: number[] | FieldValue
updatedAt?: Date
  }) {
    this.content = content
this.numbers = numbers
this.updatedAt = updatedAt
  }


  readonly content?: string

readonly numbers?: number[] | FieldValue

readonly updatedAt?: Date

  toJson(): Record<string, unknown> {
  const json: Record<string, unknown> = {}
  if (this.content != undefined) {
  json['content'] = this.content
}
if (this.numbers != undefined) {
  json['numbers'] = this.numbers
}
if (this.updatedAt != undefined) {
  json['updatedAt'] = this.updatedAt
}

  return json
}

}

export class DeleteAppUserPost {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the appUserPosts collection for reading.
 */
export const readAppUserPostCollectionReference = (
  {
  appUserId
}: {
  appUserId: string
}
): FirebaseFirestore.CollectionReference<ReadAppUserPost>

=> { return db
.collection('appUsers').doc(appUserId).collection('appUserPosts').withConverter<ReadAppUserPost>({
  fromFirestore: (ds: FirebaseFirestore.DocumentSnapshot): ReadAppUserPost => {
    return ReadAppUserPost.fromDocumentSnapshot(ds)
  },
  toFirestore: () => {
      throw new Error(`toFirestore is not implemented for ReadAppUserPost`)
  }
})
}

/**
 * Provides a reference to a appUserPost document for reading.
 * @param appUserPostId - The ID of the appUserPost document to read.
 */
export const readAppUserPostDocumentReference = ({
  appUserId,
  appUserPostId,
}: {
  appUserId: string,
  appUserPostId: string
}): FirebaseFirestore.DocumentReference<ReadAppUserPost> =>
    readAppUserPostCollectionReference({
      appUserId
    }).doc(appUserPostId);


/**
 * Provides a reference to the appUserPosts collection for creating.
 */
export const createAppUserPostCollectionReference = (
  {
  appUserId
}: {
  appUserId: string
}
): FirebaseFirestore.CollectionReference<CreateAppUserPost>

=> { return db
.collection('appUsers').doc(appUserId).collection('appUserPosts').withConverter<CreateAppUserPost>({
  fromFirestore: () => {
    throw new Error(`fromFirestore is not implemented for CreateAppUserPost`)
  },
  toFirestore: (obj: CreateAppUserPost): FirebaseFirestore.DocumentData => {
      return obj.toJson()
  }
})
}

/**
 * Provides a reference to a appUserPost document for creating.
 * @param appUserPostId - The ID of the appUserPost document to read.
 */
export const createAppUserPostDocumentReference = ({
  appUserId,
  appUserPostId,
}: {
  appUserId: string,
  appUserPostId: string
}): FirebaseFirestore.DocumentReference<CreateAppUserPost> =>
    createAppUserPostCollectionReference({
      appUserId
    }).doc(appUserPostId);


/**
 * Provides a reference to the appUserPosts collection for updating.
 */
export const updateAppUserPostCollectionReference = (
  {
  appUserId
}: {
  appUserId: string
}
): FirebaseFirestore.CollectionReference<UpdateAppUserPost>

=> { return db
.collection('appUsers').doc(appUserId).collection('appUserPosts').withConverter<UpdateAppUserPost>({
  fromFirestore: () => {
    throw new Error(`fromFirestore is not implemented for UpdateAppUserPost`)
  },
  toFirestore: (obj: UpdateAppUserPost): FirebaseFirestore.DocumentData => {
      return obj.toJson()
  }
})
}

/**
 * Provides a reference to a appUserPost document for updating.
 * @param appUserPostId - The ID of the appUserPost document to read.
 */
export const updateAppUserPostDocumentReference = ({
  appUserId,
  appUserPostId,
}: {
  appUserId: string,
  appUserPostId: string
}): FirebaseFirestore.DocumentReference<UpdateAppUserPost> =>
    updateAppUserPostCollectionReference({
      appUserId
    }).doc(appUserPostId);


/**
 * Provides a reference to the appUserPosts collection for deleting.
 */
export const deleteAppUserPostCollectionReference = (
  {
  appUserId
}: {
  appUserId: string
}
): FirebaseFirestore.CollectionReference<DeleteAppUserPost>

=> { return db
.collection('appUsers').doc(appUserId).collection('appUserPosts').withConverter<DeleteAppUserPost>({
  fromFirestore: () => {
    throw new Error(`fromFirestore is not implemented for DeleteAppUserPost`)
  },
  toFirestore: (): FirebaseFirestore.DocumentData => {
    throw new Error(`toFirestore is not implemented for DeleteAppUserPost`)
  }
})
}

/**
 * Provides a reference to a appUserPost document for deleting.
 * @param appUserPostId - The ID of the appUserPost document to read.
 */
export const deleteAppUserPostDocumentReference = ({
  appUserId,
  appUserPostId,
}: {
  appUserId: string,
  appUserPostId: string
}): FirebaseFirestore.DocumentReference<DeleteAppUserPost> =>
    deleteAppUserPostCollectionReference({
      appUserId
    }).doc(appUserPostId);



/**
 * Manages queries against the appUserPosts collection.
 */
export class AppUserPostQuery {
  /**
   * Fetches appUserPost documents.
   * @param queryBuilder - Function to modify the query.
   * @param compare - Function to sort the results.
   */
  async fetchDocuments({
      appUserId,
      queryBuilder,
      compare
  }: {
      appUserId: string,
      queryBuilder?: (
          query: FirebaseFirestore.Query<ReadAppUserPost>
      ) => FirebaseFirestore.Query<ReadAppUserPost>
      compare?: (lhs: ReadAppUserPost, rhs: ReadAppUserPost) => number
  }): Promise<ReadAppUserPost[]> {
      let query: FirebaseFirestore.Query<ReadAppUserPost> =
          readAppUserPostCollectionReference({appUserId})
      if (queryBuilder != undefined) {
          query = queryBuilder(query)
      }
      const qs: QuerySnapshot<ReadAppUserPost> = await query.get()
      let result = qs.docs.map((qds: QueryDocumentSnapshot<ReadAppUserPost>) =>
          qds.data()
      )
      if (compare != undefined) {
          result = result.sort(compare)
      }
      return result
  }

  /**
   * Fetches a specific appUserPost document.
   * @param appUserPostId - The ID of the appUserPost document to fetch.
   */
  async fetchDocument({
    appUserId,
    appUserPostId
  }: {
    appUserId: string,
    appUserPostId: string
  }): Promise<ReadAppUserPost | undefined> {
    const ds = await readAppUserPostDocumentReference({
      appUserId,
      appUserPostId
    }).get()
    return ds.data()
  }

  /**
   * Adds a appUserPost document.
   * @param CreateAppUserPost - The appUserPost details to add.
   */
  async add({
    appUserId,
    CreateAppUserPost
  }: {
    appUserId: string,
    CreateAppUserPost: CreateAppUserPost
  }): Promise<DocumentReference<CreateAppUserPost>> {
    return createAppUserPostCollectionReference({appUserId}).add(CreateAppUserPost)
  }

  /**
   * Sets a appUserPost document.
   * @param appUserPostId - The ID of the appUserPost document to set.
   * @param CreateAppUserPost - The appUserPost details to set.
   * @param options - Options for the set operation.
   */
  async set({
      appUserId,
      appUserPostId,
      CreateAppUserPost,
      options
  }: {
      appUserId: string,
      appUserPostId: string
      CreateAppUserPost: CreateAppUserPost
      options?: FirebaseFirestore.SetOptions
  }): Promise<WriteResult> {
      if (options == undefined) {
          return createAppUserPostDocumentReference({
            appUserId,
            appUserPostId
          }).set(CreateAppUserPost)
      } else {
          return createAppUserPostDocumentReference({ 
            appUserId,
            appUserPostId 
            }).set(CreateAppUserPost, options)
      }
  }

  /**
   * Updates a specific appUserPost document.
   * @param appUserPostId - The ID of the appUserPost document to update.
   * @param UpdateAppUserPost - The details for updating the appUserPost.
   */
  async update({
    appUserId,
    appUserPostId,
    UpdateAppUserPost
  }: {
    appUserId: string,
    appUserPostId: string
    UpdateAppUserPost: UpdateAppUserPost
  }): Promise<WriteResult> {
      return updateAppUserPostDocumentReference({ 
        appUserId,
        appUserPostId 
      }).update(UpdateAppUserPost.toJson())
  }

  /**
   * Deletes a specific appUserPost document.
   * @param appUserPostId - The ID of the appUserPost document to delete.
   */
  async delete({
    appUserId,
    appUserPostId
  }: {
    appUserId: string,
    appUserPostId: string
  }): Promise<WriteResult> {
      return deleteAppUserPostDocumentReference({ 
        appUserId,
        appUserPostId 
      }).delete()
  }
}
